### DISCUSSION TITLE
Deployment & Migration: 03 - Oplog

### DISCUSSION THREAD

**Discussion by alexbrowng** [directus/directus#7396] [Feature Requests]
_This is a part of a series of proposals related to create a migration system, in which I would like to know your opinion and suggestions._

_My intention is, if it is interesting for the community and the internal team gives the go-ahead, I would start developing it myself._

_Sorry if I have duplicated another idea._

_Thanks in advance._

### Needs

Create a system that allows to recreate the data model, roles and permissions in another environment allowing the maintenance and migration of the data schema.

### Use Cases

- Migrate data schema without losing data.
- Version control of the data model.
- Create a data schema for a base project.

### Background

My workflow consists of three environments, develop (local), sandbox (staging) and live (production), when I start a project I create the database, run the directus migrations and start the app to create the data model, roles and permissions, when I'm done I create a dump of the database and promote it through the different environments, when I change the data model, roles or permissions, I create a file for each collection (directus_collections, directus_fields, directus_relations, directus_roles, directus_permissions) in JSON with the fields sorted for each of the environments and do a diff to match the environments, then, with that difference I access the new environment and make the changes manually, a slow and delicate process. 😓

### Proposal

After reviewing the CollectionService and FieldServices I realized that all changes in the data model from the App or from the API call the same requests, therefore, you can create an oplog like the mongo replica or redis integrity system that generates a list with all the operations in the data model, roles and permissions to be able to execute in that same order in another system and recreate all the steps, the operations are stored in database and can be exported in JSON to keep a version control or import in another database.

### Minimal version

The initial version must have a command to create an oplog from an existing data model and a service that records each of the subsequent operations in the services: collections, fields, relations, roles and permissions, it must also allow exporting the list of operations in JSON and importing by executing the pending operations. In addition, it must implement a system of transformations that allows to update on the fly the operations when a field of the system tables changes, for example, if I create an oplog for the version X.X.X.X for a collection and I import it in a later version Y.Y.Y that includes a new field I transform the data to add the default value of that field.

**Data model**

directus_oplog (id, timestamp, version (directus version), collection, operation, data, hash (hash of data))

**Oplog Service**

- Create new log from an existing database.
- Create a operation: Add a log to the database.
- Execute a operation: Execute a log calling the service and passing the information.
- Export. Create a file with the list of operations.
- Import. Execute a file with the list of operations.
    - Check the history, only execute if it is not corrupted (the current oplog has not created any record different from the previous one).
    - Execute the records that are not in the oplog.

**Migration service**

POST /migration {"url": "https://directus.example.es", "secret": "ABC..."}: Send the oplog to another system to migrate the data.

PUT /migration [{}...]: Receives the oplog to migrate the data model. The request is signed with a secret key.

**ENV**

MIGRATION_SECRET = 'ABC...'

**App**

Page to preview the oplog history

### Futurible

- Register environments to synchronize and be able to see when is the last time I have synchronized data with another environment and if there are logs to be synchronized.
- Activate/deactivate a mode
- Create database backup
- Restore backup if migration fails

**Comment by Claratj** [comment]
That would be super interesting!
I hope the initiative goes ahead

**Comment by sandra-laguna** [comment]
I find it very interesting. I will apply it to my projects.
Thank you very much for the proposal!

**Comment by rijkvanzanten** [comment]
> After reviewing the CollectionService and FieldServices I realized that all changes in the data model from the App or from the API call the same requests, therefore, you can create an oplog like the mongo replica or redis integrity system that generates a list with all the operations in the data model

There's one big missing piece here: Database mirroring.

The user can also add in tables/columns directly in the database, which Directus normally picks up and treats as normal collections/fields. We'll have to make sure we include a way for those to be pulled in as well.

Regarding all this, have you seen https://github.com/directus/directus/discussions/3891? There's a pretty long thread discussing potential ways to do this schema migration 🙂  (though your proposal is by far the most detailed ✨ )

**Reply by alexbrowng** [reply]
I understand that when directus detects the database tables and fields it somehow adds them to the directus_collections, directus_fields and directus_relations tables at that point they would be included in the oplog. Is this the case?

**Reply by alexbrowng** [reply]
I have read the current discussion, but I think it mixes two different topics, on the one hand, the definition of the data model in files, which, I agree is a great idea but due to the large number of options I see difficult to implement in a first phase, on the other hand, the migration between environments, in some examples I have seen that it includes not only the data model but also the content of the tables, the solution I propose is the least invasive, based on models that already exists, it does not solve all the problems but I think it is a good starting point.

**Reply by rijkvanzanten** [reply]
> I understand that when directus detects the database tables and fields it somehow adds them to the directus_collections, directus_fields and directus_relations tables at that point they would be included in the oplog. Is this the case?

Nope, it doesn't! It reads it straight from the database schema and serves it as is. collections/fields is used as optional metadata to configure more specific functionality (like interface used, form layout etc)

**Reply by rijkvanzanten** [reply]
> I have read the current discussion, but I think it mixes two different topics, on the one hand, the definition of the data model in files, which, I agree is a great idea but due to the large number of options I see difficult to implement in a first phase, on the other hand, the migration between environments, in some examples I have seen that it includes not only the data model but also the content of the tables, the solution I propose is the least invasive, based on models that already exists, it does not solve all the problems but I think it is a good starting point.

Agreed! From what I can tell, there's two models that keep popping up:

1) Keep track of incremental changes, and re-apply them in a different environment
2) Have a single source of truth file schema, and use that to generate and execute a diff between the old and new version of the file

To be fair, we don't have to choose one over the other necessarily. Other services, like Hasura, chose to go with a hybrid model where (1) is used for the database migrations, and (2) is used to create/restore "dumps" of the metadata information (in our case the information in directus_collections, fields, and relations)

**Reply by rijkvanzanten** [reply]
One thing to note: As soon as you go to data-model in Directus to change anything in any table or field that isn't in directus_* tables yet, we will add it, so _maybe_ it's fine to say that we don't track changes in tables that are "Database Only"

**Reply by alexbrowng** [reply]
> > I understand that when directus detects the database tables and fields it somehow adds them to the directus_collections, directus_fields and directus_relations tables at that point they would be included in the oplog. Is this the case?
> 
> Nope, it doesn't! It reads it straight from the database schema and serves it as is. collections/fields is used as optional metadata to configure more specific functionality (like interface used, form layout etc)
> 
> 

In this case, when reading the schema you will have to create the collection, fields and relations records and detect when they are created, modified or deleted to record the changes.

Although, I understand that, if you only use directus as data mirroring the migrations are done directly in the database, I think you have to offer different solutions for different uses, my proposal is useful if you manage your data schema with directus.

**Reply by rijkvanzanten** [reply]
Yeah I can get behind that. If you're the sort of power user that relies on the direct database mirroring, it's very likely you're already using other tools to setup the direct-in-db schema, and therefore might not care about Directus' migration capabilities

**Reply by benhaynes** [reply]
That almost seems like a Docs issue, and maybe a benefit in case you _want_ some tables (temp/raw/external) to not be tracked.

Also, we could add a global project setting for `Track Direct-to-Database Changes`. Then, whenever Directus "discovers" a new database table/column, it could automatically generate the vanilla record in `directus_collections` and/or `directus_fields`. This could then be combined with the [Task Manager](https://github.com/directus/directus/discussions/3062) to actively _check_ the database for changes on a predefined interval.

Considering all of the above, could we then use our internal revision log (`directus_activity` + `directus_revisions`) to generate the diffs/migrations?

**Reply by benhaynes** [reply]
Although a different topic, managing all migrations through our internal revisions would potentially enable optional **content** migrations too. It has been requested by users to have a way to also migrate their content between environments. Seems nice to combine the logic for schema+content in one similar flow.

**Reply by rijkvanzanten** [reply]
> Considering all of the above, could we then use our internal revision log (directus_activity + directus_revisions) to generate the diffs/migrations?

It seems like the structure of this "oplog" table would have quite a bit of overlap with directus_revisions, so we might be able to use revisions for both

**Reply by alexbrowng** [reply]
The first approach I had with the migration system was to use the revisions, but, the truth is, at least in my case, when I want to migrate from one environment to another is because I have changed the data model, because, the content itself (the data) is not kept synchronized by environments, that is, I want that if I add a new field or a new option appears in the new environment but I do not want to delete or modify the content, just for that reason I wanted to separate it, also, I think that, the important thing to ensure that nothing is broken is, that the oplog has the same history of operations, because if not, the system becomes complicated, that is, if I have my local environment, I add a field and in the productive environment I add another field, it should not allow me to synchronize the oplog, because those are very complex operations, at least for a first version, the easiest to start would be to approach it as a system similar to master/slave, where I can only write in an environment and synchronize it in the rest with the same order.

**Reply by rijkvanzanten** [reply]
Also regarding content migrations @benhaynes keep in mind that the order of operations is only really important for the database schema itself (eg adding/renaming/removing columns), while the content is a more "dump and replace" sort of operation, which might not need any of this complexity

**Reply by benhaynes** [reply]
> while the content is a more "dump and replace" sort of operation, which might not need any of this complexity

True! I guess for content we'd just pull whatever's in the db directly. 👍

**Comment by dstoyanoff** [comment]
Great suggestion. I like the idea. I would just add a few more items that are nice to haves and we use them in our project:
- directus_preset - creates a predefined layout that I would like to have synced across environments
- directus_webhooks/directus_settings - would also be an useful addition

We run a whitelabeling solution, where we can have more multiple instances with the exact same schema of the tables, but different settings and branding (logo, colors, webhooks). Would be useful to have this level of configuration.

**Reply by rijkvanzanten** [reply]
@dstoyanoff In your particular use case, do you want to sync updates to the model across the fleet of instances, or is it more a "one time setup" after which the schemas diverge?

**Reply by dstoyanoff** [reply]
The flow looks like this:
- create first version of the schema
- create X instances with different branding/webhooks, rest is the same
- evolve schema
- create another X instances with different branding/webhooks, rest is the same

_Note:
By instance I mean branding instance. Then every branded instances has 2 environments (UAT/PROD)_

I think this describes it well. Basically the branding is mostly one-time setup. Changes are not uncommon and it would be nice to have, but definitely not a must as it would be bearable to be done manually.

**Reply by rijkvanzanten** [reply]
Is there a case where you evolve the schema, and then retroactively update all old created instances to the new base schema?

**Reply by dstoyanoff** [reply]
Yes, sure. Imagine it like this: you evolve the product all the time and new clients can come anytime, which means you have to plug them into the pipeline and when you evolve the product, everyone should leverage it.

Of course, this multi-branding setup is kind of specific to my usecase, so I don't really expect to be fully supported. Db migrations would be a nice start

**Comment by dstoyanoff** [comment]
Hello guys,
I've been playing around with the migrations in the last couple of days and eventually published an NPM package that handles automatic generation of migrations. It basically does this by having a "state" file that sits in a git repo. You can then create configurations from the Directus UI and run it again to analyze the diff and generate raw queries to update the state. 

I've only tested it with Postgres as this is the DB Client that my target project is using. Would appreciate if someone is able to test other databases.

Any feedback from the Directus maintaining team would also be useful to improve the tool.

https://github.com/dstoyanoff/directus-auto-migrate

**Reply by keesvanbemmel** [reply]
This is amazing man. Just tried it with my MySQL directus instance and it works! Not from the npx command though (keeps complaining about the mysql module not being found, which I can't seem to resolve).

This works:
1. Clone your repo
2. `yarn add mysql`
3. `yarn migrate my-temp-migration`

Of course with the correct .env in place.

Thanks a lot!! Will keep testing this for sure.

**Reply by dstoyanoff** [reply]
Thanks! I am glad it works well. I'd appreciate if you can give me a sign when you confirm mysql works, so I can write it in the readme. Thanks!

**Reply by dstoyanoff** [reply]
Btw, just published 0.2.1 with mysql dependency added. This should probably fix your problem.

**Reply by keesvanbemmel** [reply]
Tried it, works without adding my sql. The steps are nou:

1. Clone repo
2. `yarn install`
3. `yarn migrate temp-migration`

The npx command still complains about not being able to find mysql though.

**Reply by dstoyanoff** [reply]
Strangely it has no issues with pg and npx..

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
